### ISSUE TITLE
Editor "PCK mode" - for user-friendly mod creation, DLC creation, layered game development

### ISSUE THREAD

**Issue by trollusk** [godotengine/godot-proposals#6788]
### Describe the project you are working on

3D ARPG

### Describe the problem or limitation you are having in your project

Godot games can be modded via PCKs. However, support for this from the user/modder side is poor. The editor can't do anything with PCK files directly, you have to download a third-party tool to extract a Godot project from them. This project can then be edited, but it is a partial "skeleton" that lacks any info about the underlying game - the modder just has to know about the game structure somehow, and create a PCK that will be compatible with the game without any help from the editor.

It also makes it difficult for game creators to make DLC for games, for the same reason. Currently the game author has to make the DLC as an entirely separate project with the same structure as the main game but no access to its data/assets. 

The Bethesda games have a modding setup that has been extremely successful. The mod editor (Creation Kit) is also the tool used by Bethesda to make the games. On startup the editor loads all game data (the equivalent of loading all exported/baseline PCK files). The user then uses the editor to add or change game objects, assets, scripts etc. This feels like they are editing the game itself, but the changes are not actually made to the game's files, but are saved as a "diff", in the equivalent of a new PCK. When the game is actually run, the game loads these new PCKs after the baseline ones, in a defined order, so that the changes in the "mod" PCKs override the contents of the baseline PCKs where there is a conflict.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

- Allow the Godot editor to open and edit PCK files directly (maybe extracting them to a temp folder).
- When a PCK file is opened for editing within a Godot project, the project should be loaded first, then the PCK should be loaded as an "overlay" (its content overrides the project content).
- PCK and project content are presented as unified in the editor - the user should be able to freely inspect and edit the union of the project and PCK.
- However, the user's edits do not alter the underlying project - they only alter the PCK.
- Finally, define a way for a game to automatically load all PCKs from a folder, with a defined load order.

This feature would:
- allow game authors to easily make DLC, because they can work with the existing project while doing so.
- allow mod authors to easily make mods for the same reason
- introduce mod authors to the Godot engine, increasing the likelihood they will use it if they become game developers
- allow game authors to develop new content in a "layered" manner, keeping it separate from existing content so they can easily see what has changed if something goes wrong etc. Once the new content is finished they can de-PCKify it and merge it into the main project, or keep it as a PCK.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

File > Open PCK
File > Create new PCK
File > Close PCK

While in "PCK mode" (a PCK file is open):
- Indicate this in an obvious way (eg show the FileSystem file and directory names in a different colour)
- the FileSystem is the union of the project and PCK filesystems
- To retrieve a file within the project, the editor loads the project version if it exists, then loads the PCK version "over" it. This way any edits that are not specified in the PCK, are inherited from the base version.
- To save a change to a file, or a new file: the file is saved to the PCK. No changes are made to the underlying project files. If possible a minimal "diff" version of the file is saved to the PCK, containing only the edited properties.

Project Settings > PCKs tab
This contains an ordered list of PCK files, similar to Autoloads tab.
PCKs should reside in a defined folder or folders, which can be outside the project filesystem.
All PCKs found in those folders should automatically appear on this list (where they can be reordered, disabled etc).

If there are PCKs in this list, then:
- when the project is loaded in the editor, each PCK is loaded afterwards, in order.
- changes in later PCKs override earlier PCKs.
- the project is always edited in PCK mode. The active PCK is the last one on the list.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this would be significant new editor and engine functionality .

### Is there a reason why this should be core and not an add-on in the asset library?

Modding is more and more important to the success and longevity of games.

The Godot engine is tantalisingly close to being able to provide excellent modding facilities, but it is not there because of the rudimentary nature of PCK support.

This proposal would also make it much easier for game authors to make DLC, and to use a "layered" approach to adding new content to a game (initially create the content as a PCK, then when happy with it, de-PCKify it and merge it into the project).

Being able to advertise Godot as a game engine which can also be used as a Creation Kit-like mod creation tool for all created games, would provide it with a significant point of difference vs Unity and Unreal.

I think the broad array of benefits is compelling enough that this should be a core feature.

**Comment by KoBeWi** [comment]
> It also makes it difficult for game creators to make DLC for games, for the same reason. Currently the game author has to make the DLC as an entirely separate project with the same structure as the main game but no access to its data/assets.

Not really. You can keep DLC together with the main project, but exclude it from export (using EditorExportPlugin for most flexibility). The PCK with DLC can be created using editor plugin.

**Comment by trollusk** [comment]
This feature could be presented as "layers". PhotoShop and 3D modelling tools have layers. Godot scenes have layers. Why not layers for computer programs themselves?

**Comment by Exerionius** [comment]
> Not really. You can keep DLC together with the main project, but exclude it from export (using EditorExportPlugin for most flexibility). The PCK with DLC can be created using editor plugin.
How do you do that? Is there any examples?

**Comment by amelkor** [comment]
I think would be useful in not only DLC and modding. In my case I have 300k+ assets in the project which form the game's map. As they're complete, there's not really a need to keep them in the editor as source files, and having such amount of assets makes the editor extremely slow.
I tried to look into packing all the assets into dedicated pck files and load them in the editor, while keeping only the core assets and programming the logic, but as I understand it's not supported. In such way huge projects could be split into separate projects to make programming and asset management much easier.

**Comment by voidexp** [comment]
Currently working on a project where the game itself is a mod, mostly with content, which loads first. The idea is to open-source the core codebase, while distributing the whole game as a bundle which includes the proprietary mod. This proposal, if implemented, would infinitely aid in this scenario, as I want players to be able to open the project in Godot, load the proprietary mod and then develop their mods on top of it.

**Comment by Chaosed0** [comment]
This feels necessary for supporting mods that add scenes to the game. If a mod PCK wants to reference an existing font resource, for example, there is simply no way to do that through the editor. It seems that some code from the modder would be needed to load that font resource and apply it to the node at runtime, which removes most of the advantage of having the editor in the first place.

**Comment by voidexp** [comment]
Additionally, seems an option is needed to use only resource file paths rather than UIDs, to avoid collisions in mods developed independently.
